src/System.Net.Http.Json/src/System/Net/Http/Json/JsonContent.cs

```diff
@@ -0,0 +1,109 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+// See the LICENSE file in the project root for more information.
+
+using System.Diagnostics;
+using System.IO;
+using System.Net.Http.Headers;
+using System.Text;
+using System.Text.Json;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Http.Json
+{
+    public partial class JsonContent : HttpContent
+    {
+        internal const string JsonMediaType = "application/json";
+        internal static readonly MediaTypeHeaderValue DefaultMediaType = MediaTypeHeaderValue.Parse(string.Format("{0} {1}", JsonMediaType, Encoding.UTF8.WebName));
+
+        private readonly JsonSerializerOptions? _jsonSerializerOptions;
+        public Type ObjectType { get; }
+        public object? Value { get; }
+
+        private JsonContent(object? value, Type inputType, MediaTypeHeaderValue mediaType, JsonSerializerOptions? options)
+        {
+            if (mediaType == null)
+            {
+                throw new ArgumentNullException(nameof(mediaType));
+            }
+
+            if (inputType == null)
+            {
+                throw new ArgumentNullException(nameof(inputType));
+            }
+
+            Value = value;
+            ObjectType = inputType;
+            Headers.ContentType = mediaType;
+            _jsonSerializerOptions = options;
+        }
+
+        public static JsonContent Create<T>(T value, MediaTypeHeaderValue mediaType, JsonSerializerOptions? options = null)
+        {
+            return Create(value, typeof(T), mediaType, options);
+        }
+
+        public static JsonContent Create(object? inputValue, Type inputType, MediaTypeHeaderValue mediaType, JsonSerializerOptions? options = null)
+        {
+            if (mediaType == null)
+            {
+                throw new ArgumentNullException(nameof(mediaType));
+            }
+
+            return new JsonContent(inputValue, inputType, mediaType, options);
+        }
+
+        protected override Task SerializeToStreamAsync(Stream stream, TransportContext context)
+            => JsonSerializer.SerializeAsync(GetStreamToWriteTo(stream), Value, ObjectType, _jsonSerializerOptions);
+
+        protected override bool TryComputeLength(out long length)
+        {
+            length = 0;
+            return false;
+        }
+
+        private Stream GetStreamToWriteTo(Stream targetStream)
+        {
+            Stream jsonStream = targetStream;
+            Encoding? targetEncoding = GetEncoding(Headers.ContentType.CharSet);
+
+            // Wrap provided stream into a transcoding stream that buffers the data transcoded from utf-8 to the targetEncoding.
+            if (targetEncoding != null && targetEncoding != Encoding.UTF8)
+            {
+                jsonStream = new TranscodingWriteStream(jsonStream, targetEncoding);
```

Comment:
This TranscodingWriteStream is renting arrays from the pool, and those arrays are then returned as part of its Dispose.  But the returned Stream isn't being disposed in SerializeToStreamAsync, which means we're taking buffers out of the pool and never returning them.